Add support for SAMD20 Xplained Pro Board

sknsean opened the pull request:
I'm in the process of porting Zephyr to a custom board, MCU ATSAMD20E18.
First step is to get samd20 support included :-)

sknsean committed:
ext: Import Atmel SAMD20 header files from ASF library

Atmel Software Framework (ASF) provides a set of low-level header
files that give access to different hardware peripherals of Atmel's
ICs.

Origin: Atmel SAMD20 Series Device Support (1.2.91)
License: Apache-2.0
URL: http://packs.download.atmel.com/Atmel.SAMD20_DFP.1.2.91.atpack
Purpose: Introduction of ASF for the SAM0 series.
Maintained-by: External

Signed-off-by: Sean Nyekjaer <sean@nyekjaer.dk>

pfalcon commented:
@nzmichaelh : FYI

pfalcon commented:
@jpfaff : FYI

nzmichaelh commented:
Thanks @pfalcon, LGTM.

mnkp commented:
As far as I understand Zephyr development process no single commit should break the build. This means that commits:
- ext: atmel: rename samd folder to samd21
- boards: arduino_zero: use samd21 soc series
- arch: arm: soc: atmel_sam0: rename samd folder to samd21

should be combined into one: "arch: atmel_sam0: rename samd SoC series to samd21. While files in different folders need to be touched this is a one atomic change that we are doing.

mnkp commented on ext/hal/atmel/asf/sam0/include/samd20/sam.h:
```diff
@@ -0,0 +1,70 @@
+/**
```
This file is not required, we could remove it from commit -> less clutter. Other existing imports from Atmel ASF do not include this file.

mnkp commented on ext/hal/atmel/asf/sam0/include/samd20/samd20.h:
```diff
@@ -0,0 +1,78 @@
+/**
```
This file is not required.

mnkp commented on ext/hal/atmel/asf/sam0/include/samd20/system_samd20.h:
```diff
@@ -0,0 +1,48 @@
+/**
```
This file is not required.

mnkp reviewed (changes requested):
ASF library files have windows style endings "\r\n", we should convert it to Unix "\n". As far as I know checkpatch does not run on ext/ folder so such errors are not caught automatically. We probably should fix checkpatch.

sknsean commented:
Squashed the 3 move commits
Fixed line endings :-)

mnkp commented:
Thanks! LGTM, with a minor remark. The comment to the first commit "Make for room for other samd series" could be improved. Maybe something like "Atmel SAMD21 series was classified too broadly as SAMD. This patch names it correctly to make room for other members of SAMD series".

sknsean commented:
@mnkp agreed fixed commit msg :-)

mnkp commented:
Sorry, I was not clear enough. The commit's first line ""arch: atmel_sam0: rename samd SoC series to samd21" was OK. It's short, to the point and said what the commit did. Only the body of the commit message was not very descriptive. I guess you know already that the first comment line in git is special. To sum it up I would write a commit message like this

arch: atmel_sam0: rename samd SoC series to samd21

Atmel SAMD21 series was classified too broadly as SAMD. This patch names it correctly to make room for other SAMD series.

sknsean commented:
Okay :-)
I removed the first line as it changed files in a lot of other directories than arch/atmel_sam0..
I'll re-add that line and fix some issues after the spi driver was merged.

sknsean commented:
@mnkp fixed and green again :-)

sknsean commented:
@nashif @galak What do i need to do to get this merged?
I'm using this for my own custom board, do I need to buy the evalkit with the CPU and make a complete PR for support for that?

galak commented:
> @nashif @galak What do i need to do to get this merged?
> I'm using this for my own custom board, do I need to buy the evalkit with the CPU and make a complete PR for support for that?

We need the code to be testable if we accept it upstream, so having a CPU/SoC & board port that uses the code is something of a prereq for merging.  I'm find with merging the rename patch right now if you want to pull that out into its own PR.

sknsean commented:
Great will buy the evalkit :-)
And move the rename patch to it's own PR

galak commented:
I'm good with this, but will wait on SoC & board port before accepting.

sknsean committed:
boards: add board and DTS definitions for the SAMD20 Xplained Pro Board

Complete board specifications for the SAMD20 Xplained Board.

Signed-off-by: Sean Nyekjaer <sean@nyekjaer.dk>

sknsean committed:
arch: add support SAMD20 used in the SAMD20 Xplained Pro Board

This add basic support the SAMD20 inkl. SPI and UART

Signed-off-by: Sean Nyekjaer <sean@nyekjaer.dk>

sknsean committed:
dts: arm: sam0: move contents samd21 to samd

samd21.dtsi includes now the same content need for SAMD20,
move it to samd.dtsi and include it from samd21.dtsi.
Then later USB support can be added to the samd21.dtsi seperatly
from the samd20 etc.

Signed-off-by: Sean Nyekjaer <sean@nyekjaer.dk>

sknsean committed:
arch: atmel_sam0: move clk config options to common Kconfig

The SAMD0 Family have almost the same clock options.
The samd20 and samd21 only clocks is not implemented yet.

Signed-off-by: Sean Nyekjaer <sean@nyekjaer.dk>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

sknsean commented:
@galak and @nzmichaelh
Please take a look :-)

nzmichaelh commented on dts/arm/atmel/samd.dtsi:
```diff
@@ -0,0 +1,137 @@
+/*
```
nit: Atmel now have a SAMD51 series which seems to be a SAM0 with a Cortex M4.  I wouldn't worry about it now, but we may need to rename this in the future.

mnkp replied:
We probably shouldn't be doing it. SAMD family has the following series: D09, D10, D11, D20, D21, DA1, HA1 and they are all quite different. E.g several D10 peripherals are located at different addresses than D21/D20. While it's true that D20 and D21 look fairly similar if we try to maintain one common file it will very quickly get messy. Even at present this file contains an entry for USB0 peripheral which is missing on D20 chip.

Let's maybe keep one dtsi file per series and only ensure that they are easy to compare with diff.

nzmichaelh commented on arch/arm/soc/atmel_sam0/samd20/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,40 @@
+	default samd20j17 if SOC_PART_NUMBER_SAMD20J17
+	default samd20j18 if SOC_PART_NUMBER_SAMD20J18
+
+config NUM_IRQS
```
optional: move this to the generic SAM0 configuration

nzmichaelh commented on arch/arm/soc/atmel_sam0/samd20/Kconfig.series:
```diff
@@ -0,0 +1,15 @@
+# Kconfig - Atmel SAMD MCU series
```
"Atmel SAMD20 MCU series"

nzmichaelh commented on arch/arm/soc/atmel_sam0/samd20/soc.c:
```diff
@@ -0,0 +1,196 @@
+/*
```
WDYT about making a atmel_sam0/common directory and putting this in there instead?  This is how the atmel_sam/ common code is handled.

sknsean replied:
They are not 100% identical, if we choose to add a common directory, then we would have to deal with some if defs

nzmichaelh commented on arch/arm/soc/atmel_sam0/samd20/soc_pinmap.h:
```diff
@@ -0,0 +1,36 @@
+/*
```
Note that this has moved to pinmux.c.  Consider doing the same in a follow up CL.

nzmichaelh commented on boards/arm/atsamd20_xpro/atsamd20_xpro.dts:
```diff
@@ -0,0 +1,61 @@
+
+&sercom2 {
+	status = "ok";
+	compatible = "atmel,sam-i2c-twi";
```
Huh, is this correct?  I thought the I2C port was also a SERCOM mode...

sknsean replied:
are we still missing an i2c driver for the sam0?

henrikbrixandersen replied:
@sknsean Corrrect. There is no in-tree TWI driver for the sam0 series yet.

sknsean replied:
Great one more thing on the TODO list :-)

nzmichaelh commented on boards/arm/atsamd20_xpro/support/openocd.cfg:
```diff
@@ -0,0 +1,22 @@
+transport select swd
+
+# chip name
+set CHIPNAME at91samd21e18
```
Do these need updating?

dbkinder commented on boards/arm/atsamd20_xpro/doc/atsamd20_xpro.rst:
```diff
@@ -0,0 +1,140 @@
+
+The SAM D20 Xplained Pro evaluation kit is ideal for evaluation and
+prototyping with the SAM D20 Cortex®-M0+ processor-based
+microcontrollers. Including Atmel’s Embedded Debugger (EDBG), which
```
Change ``Including`` to ``The kit includes``

dbkinder reviewed (changes requested):
One tweak, otherwise looks fine.  Thanks.

henrikbrixandersen commented on boards/arm/atsamd20_xpro/atsamd20_xpro.dts:
```diff
@@ -0,0 +1,56 @@
+		 */
+
+#if defined(CONFIG_FILE_SYSTEM_NFFS)
+		nffs_partition: partition@3c000 {
```
Please sync up with the changes in 9968cda453ac7a91d513b6a50817c926c3fe5cc6.

sknsean replied:
:+1:

henrikbrixandersen commented on boards/arm/atsamd20_xpro/doc/atsamd20_xpro.rst:
```diff
@@ -0,0 +1,140 @@
+========
+
+The SAMD20 MCU has 6 SERCOM based SPIs. On the SAM D20 Xplained Pro,
+SPI0 is available on the EXT1 connector.
```
Please sync up with be52e3c456f0694fb98a4704860102d61048fd05.

sknsean replied:
:+1:

dbkinder reviewed (approved):
+1 for doc

mnkp commented on arch/arm/soc/atmel_sam0/Kconfig:
```diff
@@ -13,6 +13,39 @@ config SOC_FAMILY
+	depends on SOC_ATMEL_SAMD_XOSC
+	bool "XOSC"
+
+config SOC_ATMEL_SAMD_OSC8M_AS_MAIN
```
Settings in this file should be applicable to the whole SAM0 family, not only SAM Dxx series. SAM0 family has following members: B11, C20, C21, D09, D10, D11, D20, D21, DA1, HA1, L21, L22, R21, R30. Some of them don't have 8 MHz RC oscillator.

sknsean replied:
Could we call it SAMD2X then?

mnkp replied:
If we are able to reduce the number of configuration files by recognizing similarities between different chips that's probably good. We would however need to ensure that our implementation is correct and there are no (not yet implemented) series that will break it. Looking at the datasheets of D09, D10, D11, D20, D21, DA1 series they all seem to have identical clock subsystem. So using SAMD common clock options may be an option. But we would need to move it to some other Kconfig file as this one is meant for the whole SAM0 family. Also we would need to ensure that if we ever add e.g. B11 series the SAMD options will not be sourced.

That said this is not how other arch/arm/soc/* families/series are configured. So far every series has its own independent configuration files. It would be best to ask @galak or @MaureenHelm what's the best way to proceed.

galak replied:
I've cleaned this up, as we see new SoCs added and new patterns we can tweak things.  for now I added a Kconfig.samd2x for that stuff.

galak commented:
Still want to try and get this merged?

sknsean commented:
Yes that’s why the PR is still open...
It seems like a lot of people had their say, but I still don’t know who the maintainer is and thereby I don’t know what needs to be fixed before it can be merged. 
The purpose of this was not to fix the bad handling/the whole world of multiple sam0 SoCs.

So please provide a list of what needs to be done

galak commented:
> Yes that’s why the PR is still open...
> It seems like a lot of people had their say, but I still don’t know who the maintainer is and thereby I don’t know what needs to be fixed before it can be merged.
> The purpose of this was not to fix the bad handling/the whole world of multiple sam0 SoCs.
> 
> So please provide a list of what needs to be done

I merged down some of the commits, and updating a few things to sync with the latest upstream changes.  Will review to see if I have any other comments, while it runs CI checks.

skny-prevas commented:
Great, thanks :+1:

sknsean commented:
Great thanks :+1:
